File: demo_app/app/src/main/java/com/bandyer/demo_android_sdk/utils/networking/MockedNetwork.java
/*
 * Copyright (C) 2019 Bandyer S.r.l. All Rights Reserved.
 * See LICENSE.txt for licensing information
 */

package com.bandyer.demo_android_sdk.utils.networking;

import android.app.Activity;
import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;

import com.bandyer.demo_android_sdk.utils.storage.ConfigurationPrefsManager;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * WARNING!!!
 * The networking package is used only to fetch the users, to make the demo app run out of the box,
 * with the least efforts.
 * <p>
 * MockedNetwork
 *
 * @author kristiyan
 */
public class MockedNetwork {

    private static Call<BandyerUsers> call;

    private static Call<Void> registerDevice;
    private static Call<Void> unregisterDevice;

    private static GetBandyerUsersCallback networkCallback;

    public interface GetBandyerUsersCallback {

        void onUsers(List<String> users);

        void onError(String error);
    }

    public static void getSampleUsers(Activity context, GetBandyerUsersCallback callback) {
        cancelGetUsers();
        networkCallback = callback;
        String apiKey = ConfigurationPrefsManager.getApiKey(context);
        String envName = ConfigurationPrefsManager.getEnvironmentName(context);

        call = APIClient.getClient(apiKey, envName).create(APIInterface.class).getUsers();
        call.enqueue(new Callback<BandyerUsers>() {
            @Override
            public void onResponse(@NonNull retrofit2.Call<BandyerUsers> call, @NonNull Response<BandyerUsers> response) {
                if (networkCallback == null) return;
                if (response.body() == null || response.body().user_id_list == null || !response.isSuccessful()) {
                    networkCallback.onError("No users found or credentials are invalid!");
                    return;
                }
                networkCallback.onUsers(response.body().user_id_list);
            }

            @Override
            public void onFailure(@NonNull retrofit2.Call<BandyerUsers> call, @NonNull Throwable t) {
                if (networkCallback == null) return;
                networkCallback.onError(t.getMessage());
            }
        });
    }

    public static void registerDeviceForPushNotification(Context context, String userAlias, String devicePushToken) {
        cancelRegisterUser();
        String appId = ConfigurationPrefsManager.getAppId(context);
        String apiKey = ConfigurationPrefsManager.getApiKey(context);
        String envName = ConfigurationPrefsManager.getEnvironmentName(context);
        String pushProvider = ConfigurationPrefsManager.getPushProvider(context);

        DeviceRegistrationInfo info = new DeviceRegistrationInfo(userAlias, appId, devicePushToken, pushProvider);
        registerDevice = APIClient.getClient(apiKey, envName).create(APIInterface.class).registerDeviceForPushNotifications(info);

        registerDevice.enqueue(new Callback<Void>() {
            @Override
            public void onResponse(@NonNull Call<Void> call, @NonNull Response<Void> response) {
                if (!response.isSuccessful()) {
                    Log.e("PushNotification", "Failed to register device for push notifications!");
                }
            }

            @Override
            public void onFailure(@NonNull Call<Void> call, @NonNull Throwable t) {
                Log.e("PushNotification", "Failed to register device for push notifications!");
            }
        });
    }

    public static void unregisterDeviceForPushNotification(Context context, String userAlias, String devicePushToken) {
        cancelUnRegisterUser();
        String appId = ConfigurationPrefsManager.getAppId(context);
        String apiKey = ConfigurationPrefsManager.getApiKey(context);
        String envName = ConfigurationPrefsManager.getEnvironmentName(context);

        unregisterDevice = APIClient.getClient(apiKey, envName).create(APIInterface.class).unregisterDeviceForPushNotifications(userAlias, appId, devicePushToken);

        unregisterDevice.enqueue(new Callback<Void>() {
            @Override
            public void onResponse(@NonNull Call<Void> call, @NonNull Response<Void> response) {
                if (!response.isSuccessful()) {
                    Log.e("PushNotification", "Failed to unregister device for push notifications!");
                }
            }

            @Override
            public void onFailure(@NonNull Call<Void> call, @NonNull Throwable t) {
                Log.e("PushNotification", "Failed to unregister device for push notifications!");
            }
        });
    }

    private static void cancelGetUsers() {
        if (call == null) return;
        if (!call.isExecuted()) call.cancel();
        call = null;
        networkCallback = null;
    }

    private static void cancelUnRegisterUser() {
        if (unregisterDevice == null) return;
        if (!unregisterDevice.isExecuted()) unregisterDevice.cancel();
        unregisterDevice = null;
    }

    private static void cancelRegisterUser() {
        if (registerDevice == null) return;
        if (!registerDevice.isExecuted()) registerDevice.cancel();
        registerDevice = null;
    }

    public static void cancel() {
        cancelGetUsers();
        cancelRegisterUser();
        cancelUnRegisterUser();
    }
}

File: app_utilities/src/main/java/com/bandyer/app_utilities/networking/MockedNetwork.kt
/*
 * Copyright (C) 2020 Bandyer S.r.l. All Rights Reserved.
 * See LICENSE.txt for licensing information
 */
package com.bandyer.app_utilities.networking

import android.app.Activity
import android.content.Context
import android.util.Log
import com.bandyer.app_configuration.external_configuration.model.PushProvider
import com.bandyer.app_utilities.storage.ConfigurationPrefsManager
import com.bandyer.app_utilities.networking.APIClient.getClient
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

/**
 * WARNING!!!
 * The networking package is used only to fetch the users, to make the demo app run out of the box,
 * with the least efforts.
 *
 *
 * MockedNetwork
 *
 * @author kristiyan
 */
object MockedNetwork {
    private var getUsers: Call<BandyerUsers?>? = null
    private var registerDevice: Call<Void?>? = null
    private var unregisterDevice: Call<Void?>? = null

    @JvmStatic
    fun getSampleUsers(context: Activity, callback: GetBandyerUsersCallback?) {
        cancelGetUsers()
        val configuration = ConfigurationPrefsManager.getConfiguration(context)
        getUsers = getClient(configuration.apiKey!!, configuration.environment)!!.create(APIInterface::class.java).users
        getUsers!!.enqueue(object : Callback<BandyerUsers?> {
            override fun onResponse(call: Call<BandyerUsers?>, response: Response<BandyerUsers?>) {
                getUsers = null
                if (response.body() == null || response.body()!!.user_id_list == null || !response.isSuccessful) {
                    callback?.onError("No users found or credentials are invalid!")
                    return
                }
                callback?.onUsers(response.body()!!.user_id_list)
            }

            override fun onFailure(call: Call<BandyerUsers?>, t: Throwable) {
                if (call.isCanceled) return
                getUsers = null
                callback?.onError(t.message)
            }
        })
    }

    @JvmStatic
    fun registerDeviceForPushNotification(userAlias: String, pushProvider: PushProvider, devicePushToken: String, apiKey: String, appId: String, environment: String, callback: Callback<Void?>?) {
        cancelRegisterUser()
        val info = DeviceRegistrationInfo(userAlias, appId, devicePushToken, pushProvider.name)
        registerDevice = getClient(apiKey, environment)!!.create(APIInterface::class.java).registerDeviceForPushNotifications(info)
        registerDevice!!.enqueue(object : Callback<Void?> {
            override fun onResponse(call: Call<Void?>, response: Response<Void?>) {
                registerDevice = null
                callback?.onResponse(call, response)
            }

            override fun onFailure(call: Call<Void?>, t: Throwable) {
                if (call.isCanceled) return
                registerDevice = null
                callback?.onFailure(call, t)
            }
        })
    }

    @JvmStatic
    fun unregisterDeviceForPushNotification(userAlias: String, devicePushToken: String, apiKey: String, appId: String, environment: String) {
        cancelUnRegisterUser()
        unregisterDevice = getClient(apiKey, environment)!!.create(APIInterface::class.java).unregisterDeviceForPushNotifications(userAlias, appId, devicePushToken)
        unregisterDevice!!.enqueue(object : Callback<Void?> {
            override fun onResponse(call: Call<Void?>, response: Response<Void?>) {
                if (call.isCanceled) return
                if (!response.isSuccessful) {
                    Log.e("PushNotification", "Failed to unregister device for push notifications!")
                }
            }

            override fun onFailure(call: Call<Void?>, t: Throwable) {
                if (call.isCanceled) return
                Log.e("PushNotification", "Failed to unregister device for push notifications!")
            }
        })
    }

    private fun cancelGetUsers() {
        getUsers?.cancel()
        getUsers = null
    }

    private fun cancelUnRegisterUser() {
        unregisterDevice?.cancel()
        unregisterDevice = null
    }

    private fun cancelRegisterUser() {
        registerDevice?.cancel()
        registerDevice = null
    }

    fun cancel() {
        cancelGetUsers()
        cancelRegisterUser()
        cancelUnRegisterUser()
    }

    interface GetBandyerUsersCallback {
        fun onUsers(users: List<String?>?)
        fun onError(error: String?)
    }
}